api/services/bulk-docs.js

```diff
@@ -26,45 +19,96 @@ const checkForDuplicates = docs => {
+  if (ids.length === 0) {
+    return Promise.resolve([]);
+  }
+  return db.medic.allDocs({ keys: ids, include_docs: true })
+    .then(extractDocs);
+};
+
+const incrementAttemptCounts = (docs, attemptCounts) => {
+  docs.forEach(doc => {
+    attemptCounts[doc._id] = (attemptCounts[doc._id] || 0) + 1;
   });
 };
 
-module.exports = {
-  bulkDelete: (docs, res, { batchSize = 100 } = {}) => {
-    let docsToDelete;
-    const keys = docs.map(doc => doc._id);
-    return db.medic.allDocs({ keys, include_docs: true })
-      .then(result => {
-        docsToDelete = markAsDeleted(result);
-        return utils.updateParentContacts(docsToDelete);
-      })
-      .then(docsToUpdate => {
-        const allDocs = docsToDelete.concat(docsToUpdate);
-        checkForDuplicates(allDocs);
-        const batches = generateBatches(allDocs, batchSize);
+const deleteDocs = (docsToDelete, deletionAttemptCounts, updateAttemptCounts, cumulativeResult = []) => {
+  const finishedModifications = cumulativeResult.map(docUpdate => docUpdate.id);
+  const parentMap = {};
+  let docsToUpdate;
+  let docsToModify;
+  docsToDelete.forEach(doc => doc._deleted = true);
+  incrementAttemptCounts(docsToDelete, deletionAttemptCounts);
+  return utils.updateParentContacts(docsToDelete, parentMap)
+    .then(updatedParents => {
+      docsToUpdate = updatedParents;
+      incrementAttemptCounts(docsToUpdate, updateAttemptCounts);
+      docsToModify = docsToDelete.concat(docsToUpdate).filter(doc => !finishedModifications.includes(doc._id));
+      return db.medic.bulkDocs(docsToModify);
+    })
+    .then(result => {
+      result = result.filter(docUpdate => docUpdate.status !== 404);
+      const deletionFailures = [];
+      const updateFailures = [];
+      const deletionIds = docsToDelete.map(doc => doc._id);
+      const updateIds = docsToUpdate.map(doc => doc._id);
+      const errorIds = result
+        .map((docUpdate, index) => docUpdate.error && (docUpdate.id || docsToModify[index]._id))
+        .filter(id => id);
+      errorIds.forEach(id => {
+        if (deletionIds.includes(id) && deletionAttemptCounts[id] < 4) {
+          deletionFailures.push(id);
+          result = result.filter(docUpdate => docUpdate.id !== id);
+        } else if (updateIds.includes(id) && updateAttemptCounts[id] < 4) {
+          updateFailures.push(id);
+          result = result.filter(docUpdate => docUpdate.id !== id);
+        }
+      });
 
-        res.type('application/json');
-        res.setHeader('X-Content-Type-Options', 'nosniff');
-        res.write('[');
+      cumulativeResult = cumulativeResult.concat(result);
+      if (deletionFailures.length > 0 || updateFailures.length > 0) {
+        return fetchDocs(deletionFailures)
+          .then(docsToDelete => {
+            const updatesToRetry = updateFailures.map(id => parentMap[id]);
+            return deleteDocs(docsToDelete.concat(updatesToRetry), deletionAttemptCounts, updateAttemptCounts, cumulativeResult);
```

Comment:
To the casual observer this looks like you're passing update ids to the `deleteDocs` function to be deleted. I'm presuming this isn't true, and is part of the machination around how your function deals with situations where a document gets successfully deleted but it's parent update fails. It's not clear from reading it though.

Reply:
Exactly right, I'll try and make things a bit clearer here